pkg/util/log/file.go

```diff
@@ -259,14 +261,18 @@ func (l *fileSink) flushAndMaybeSyncLocked(doSync bool) {
 		// recursive back-and-forth between the copy of FATAL events to
 		// OPS and disk slowness detection here. (See the implementation
 		// of logfDepth for details.)
+		sev := severity.ERROR
+		// We default to assuming a fatal on log stall.
+		if l.fatalOnLogStall == nil || l.fatalOnLogStall() {
+			sev = severity.FATAL
+			// The write stall may prevent the process from exiting. If the process
+			// won't exit, we can at least terminate all our RPC connections first.
+			//
+			// See pkg/cli.runStart for where this function is hooked up.
+			MakeProcessUnavailable()
```

Comment:
Note that this doesn't cause an exit, which is what happened in the referenced issue. There is a separate timer in `logfDepthInternal` which invokes `exitFunc`.

Reply:
Yep I'm aware. The severity being Fatal triggers that exit function, which is what I'm mostly changing here.

Reply:
It's not immediately clear why we have `ExitTimeoutOnFatalLog` and `maxSyncDuration`, which behave rather differently.

Reply:
> The severity being Fatal triggers that exit function, which is what I'm mostly changing here.

Oh, I see, but that still leaves calls like `l.Error` exposed to the same exit. Don't we want to plug that entirely via the same cluster setting?

Reply:
@srosenberg Good question and thanks for asking it as it sent me on a detour and shored up my understanding of the current logging code.

1) I don't think we want to plug all fatal logs with the cluster setting; other reasons to fatally crash the process (eg. inconsistent stores or ranges) should continue to do their thing. We just want to avoid crashing just because of disk stalls being seen by the logger.

2) It looks like `ExitTimeoutOnFatalLog` is the timeout on any fatal log entry taking too long to write to any sink (fluent, etc), while `maxSyncDuration` is for any log entry (not necessarily a fatal one) writing to a _file_ sync. Given that definition, and given point 3 below, I think having the cluster setting just neutralize the latter config parameter is correct. I'll fix the maxSyncDuration comment though to clarify this.

3) There's _another_ place in `loggerT.outputLogEntry` that actually calls the exit function, assuming the Shoutf succeeds in writing the fatal message to the ops log within the ExitTimeoutOnFatalLog. So it's still correct to just change severity to Error for disk stalls and bypass all the crashing logic.

Reply:
@itsbilal Thanks for taking a closer look and summarizing the findings! I do agree that this PR addresses the immediate issue, namely `exit 7` resulting from an expired (`maxSyncDuration`) timer in `fileSink.flushAndMaybeSyncLocked`,  which invokes `outputLogEntry` with `severity.FATAL`; reducing the severity effectively disables the exit.

I do want to mention one _pathological_ case which still remains at large. Owing to [1], it's theoretically possible for `write` to stall _before_ `flush`. (Recall that `outputLogEntry` invokes `sink.output`, which does `write`, followed by `flush`.) Now, the `flushDaemon` is scheduled to periodically invoke `flush`. Imagine the following scenario. A `write` has been issued (which may or may not end up rotating logs), subsequently the runtime stops the world long enough such that a _potential_ disk stall goes undetected; i.e., `maxSyncDuration` timer never expires yet we spend 10 seconds waiting for `write` and another 5 for STW.

[1] https://github.com/cockroachdb/cockroach/issues/56893

Reply:
@srosenberg thanks for highlighting that! We probably should reopen #56893 and try to address that on its own. It's just a lower priority issue seeing as Pebble's disk stall detection (ie. the stuff configured in `pkg/storage/pebble.go` through `WithDiskHealthChecks`) has become far more robust over the years since that issue was filed, in detecting stalls in the data directories/volumes themselves.

I remember questioning if we need to even crash if the logger is stalling but the stores aren't, so by that lower prioritization it's less of a concern if some stalls are going undetected by the logger as long as Pebble is robust at catching stalls where they can really hose the cluster.